rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  * `dir`: The subdirectory of the repository in which `flake.nix` is
+    located. This parameter enables having multiple flakes in a
+    repository. The default is the root directory.
+
+  For example, the following are valid Git flake references:
+
+  * `https://example.org/my/repo.git`
+  * `https://example.org/my/repo.git?dir=flake1`
+  * `ssh://git@github.com:NixOS/nix.git?ref=v1.2.3`
+  * `git://github.com/edolstra/dwarffs.git?ref=unstable&rev=e486d8d40e626a20e06d792db8cc5ac5aba9a5b4`
+  * `file:///home/my-user/some-repo/some-repo.git`
+
+* Local paths. These have the form
+
+      <path>(\?<params)?
+
+  where `<path>` must refer to (a subdirectory of) a Git repository.
+  These differ from `file://` flake references in a few ways:
+
+  * They refer to the working tree (unless an explicit `rev` or `ref`
+    is specified), so evaluation can access dirty files. (Dirty files
+    are files that are tracked by git but have uncommitted changes.)
+
+  * The `dir` parameter is automatically derived. For example, if
+    `/foo/bar` is a Git repository, then the flake reference
+    `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.
+
+* GitHub repositories. These are downloaded as tarball archives,
+  rather than through Git. This is often much faster and uses less
+  disk space since it doesn't require fetching the entire history of
+  the repository. On the other hand, it doesn't allow incremental
+  fetching (but full downloads are often faster than incremental
+  fetches!). The syntax is:
+
+      github:<owner>/<repo>(/<rev-or-ref>)?(\?<params>)?
+
+  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a
+  commit hash (`rev`). Note that unlike GitHub, Git allows fetching by
+  commit hash without specifying a branch or tag.
+
+  The only supported parameter is `dir` (see above).
+
+  Some examples:
+
+  * `github:edolstra/dwarffs`
+  * `github:edolstra/dwarffs/unstable`
+  * `github:edolstra/dwarffs/d3f2baba8f425779026c6ec04021b2e927f61e31`
+
+* Indirections through the flake registry. These have the form
+
+      <flake-id>(/<rev-or-ref>(/rev)?)?
+
+  These perform a lookup of `<flake-id>` in the flake registry. The
+  specified `rev` and/or `ref` are then merged with the entry in the
+  registry. (See below.) For example, `nixpkgs` and
+  `nixpkgs/release-19.03` are indirect flake references.
+
+In the future, we should also add tarball flake references
+(e.g. `https://example.org/nixpkgs.tar.xz`). It would also be
+straight-forward to add flake references to Mercurial repositories
+since Nix already has Mercurial support.
+
+## Flake registries
+
+Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
+"direct" flake references (i.e. any type of flake reference that's not
+an indirection). This is a convenience to users, allowing them to do
+
+    nix run nixpkgs:hello
+
+rather than
+
+    nix run github:NixOS/nixpkgs:hello
+
+There are multiple registries:
+
+* The global registry
```

Comment:
Why do we need a global registry at all? Could we not just, eg, provide hard-coded defaults for `nix` and `nixpkgs`, and let users do the rest? This would avoid becoming a package repository to any extent, which I think would be a very good thing.

Reply:
@alyssais what would the defaults be?

Reply:
Maybe

- `nix = https://github.com/NixOS/nix stable`
- `nixpkgs = https://github.com/NixOS/nixpkgs unstable`

Reply:
We don't *need* a registry, but it seems important for UX. In fact it's more important here than in tools like Cargo, since the registry is used in a command line interface. We don't want to require people to type `nix run github:NixOS/patchelf` instead of `nix run patchelf`. Special-casing `nixpkgs` seems ugly, and it makes every other flake second-class compared to `nixpkgs` (which is bad since we should move away from the monorepo approach since it doesn't scale).

Reply:
> which is bad since we should move away from the monorepo approach since it doesn't scale

I disagree here, which is probably also the reason I'm a bit hesitant with the flakes approach in general. For me the monorepo approach is one of the major selling points of nixpkgs. Probably *the* selling point. It lowers the barrier to contribution a lot. It makes it very easy to find any information I want with a single grep. I can bisect any problem. I can test if my update breaks any known reverse dependencies. There is a central point of trust, even if that trust is very diffused at the moment.

I imagine a bunch of single-package repos loosely tied together with lockfiles will actually scale less. Why would I want `patchelf` to be in its own flake vs. in `nixpkgs`?

Reply:
I think a global registry as proposed is a bad idea and should be avoided. That doesn't mean we can't have a local registry, much as channels are registered now. `nix` and `nixpkgs` could be registered by default in the local registry file when Nix is installed.

I'm alarmed by discussion of moving away from a monorepo; I agree that this is a major selling point of NixOS.

Reply:
I would also like to add that "special case `nixpkgs` flake in the UX" is not the only option. I see no reason the user cannot also specify the default flake that resolution should apply to, or alternatively, no default, requiring explicit flake name references. I don't think this would be surprising or unreasonable to support, and would make very obvious use cases easier to support.

For example, when I run `nix run hello`, what if I want the stable branch? Right now from my reading, `nix run nixpkgs/release-19.03:hello` would be used to do this, but if we're talking about UX vs `nix run hello`, I think I think having to do this incantation is generally worse than simply having to write `nixpkgs:hello` for the sake of flake-ambiguity. But again I feel this is a false dichotomy. Of course, I could override the flake registry file with the proposal as it stands, and change the `nixpkgs` flake to use `?ref=release-19.03`, but then I'm basically cloning the entire registry to tweak one setting (so what if it falls out of date?). What if I simply *do* want `nix run hello` to use the stable branch? What if I don't want *any* default? I don't think we should dictate project policy in this manner based on what we _think_, _might be_, a UX concern.

Of course by having user-local flake "repositories", one could argue we are re-introducing a source of in-determinism (no two users necessarily have the same local flake config). But by the other points in the design, flakes are strictly more auditable and easier to manage anyway: they more easily accounted for than `NIX_PATH` search includes or `getEnv` nonsense when debugging builds, etc. A large amount of nonsense can be traced back to stray search includes for example. Asking the user to `cat ~/.config/nix/flakes.json` when debugging something seems far superior than what we have now, global registries or not.

---

Finally, I am also wary of any discussion to move away from a monorepository. "It does not scale" is not evidently true to me in any way. Considering we are one of the most traffic'd repos on all of GitHub and other infrastructural priorities, I'd say it's actually one of our lesser concerns. The monorepo approach of Nix is a huge advantage which I consistently tell people about: it means making atomic changes across components is viable and feasible, that all source code and context (namely, `git` history) is locally available, and that we all have a singular place to coordinate and discuss (mostly). One of the costs we pay, however, and the one that upsets people (from my impressions) in our current model, is that our reliance on tools like `hackage2nix` causes huge amount of "commit object bloat" with +10,000 line diffs and large commit objects. But I am quite willing to argue that the proposed cure, in this case, is *far, far* worse than the disease could ever be.

On top of this, even assuming the Flake RFC goes through -- predicating *any* features or design decisions for Flakes on the *future intention* to "fix" the monorepo is inherently suspect to me. It has not been decided or discussed widely, and I actually suspect the pushback to such a change would be *far* greater than Flakes could ever provoke. It is not a factor that should be considered in the discussions or design of the Flake tooling or UX, as far as I'm concerned. There are plenty of other benefits to think about for now.

Reply:
One use case to consider are companies, where it could be monorepo or polyrepo.

Currently registry input is non-deterministic, so one user could have different version than the other, resulting in different lock output.

For the polyrepo case I suggest:

- registry becomes a nix file, so that it can be overriden/expanded
- adding registry field to flakes like `registry = globalRegistry: (self: super: { extraIdentifier = ...; })` with a default value of `id` function

That's an alternative way to achieve https://github.com/NixOS/rfcs/pull/49/files#r304091661

Reply:
Whatever we do with `haskellPackages`, the problem we want to solve is «checking in huge autogenerated files»; if we solve this, the main scaling problem of monorepo — deciding when to pin old dependency versions — is at least more transparent with a monorepo, and needs to be solved somehow in any layout.

Splitting the repo _and_ collecting it back via flake registry sounds like «NixOS/Nixpkgs repository split», only larger scale; we actually moved towards monorepo after the move from SVN broke explicit synchronisation of the two subrepositories.

Reply:
Thought: these problems would potentially be solved if Nixpkgs were a monorepo of flakes. Stuff like Haskell generated files could be fetched from elsewhere as a flake, but locked to a specific version.

I haven’t thought in depth yet about how this would work with lockfiles, etc.

Reply:
@alyssais, isn't that exactly what @7c6f434c mentions in their second paragraph as being problematic?

Reply:
@toonn Not exactly — the top-level repository with subtrees does provide a sequence of «cuts» with exact commit choice for all parts; this, for example, provides easier checkouts of historical states (and also that elusive state «what people are considering master branch head right now») than fully independent repos. Separate NixOS and Nixpkgs repos hoted on GitHub lacked full tooling support for this.

Reply:
Regarding monorepos vs polyrepos: flakes make a polyrepo world easier but don't mandate it. So a decision to move away from a monorepo Nixpkgs/NixOS is separate from this RFC. Ultimately however the monorepo approach fundamentally doesn't scale: it shouldn't be necessary for everybody who just wants to share a NixOS module to have to add it to the `nixpkgs` repo. Note by the way that any NixOS system is already a polyrepo configuration because a system's `configuration.nix` is not part of the nixpkgs repo. But Nix at the moment lacks a good way to manage that dependency.

Regarding registries: the UX value of global registries is obvious, IMHO, from tools like Cargo or npm.

Slightly tangentially, here is a post [1] discussing `crates.io` policies, which would be useful for the flake registry as well. It also discusses why `crates.io` doesn't have namespaced packages.

[1] https://internals.rust-lang.org/t/crates-io-package-policies/1041/1

Reply:
> Why would I want patchelf to be in its own flake vs. in nixpkgs?

That's actually a good example :-) The reason you'd want that is because the patchelf package in `nixpkgs` is several years behind the upstream package. With patchelf-as-a-flake you're no longer dependent on a single monorepo.

Reply:
> Regarding registries: the UX value of global registries is obvious, IMHO, from tools like Cargo or npm.

npm’s global registry has been a disaster. There’s been account compromises, typosquatting, name squatting in general, and useful names lie dormant because somebody grabbed them years ago.

I am strongly opposed to a global registry, especially to start with. If somebody wants to start a global registry outside of Nix to show its value, fair enough. Maybe we can adopt it. But we don’t need it for flakes, so I don’t think we should do it now.

Reply:
A search engine for flakes might be a nice idea, however. Again, though, it can be external to Nix.

Reply:
> The reason you'd want that is because the patchelf package in nixpkgs is several years behind the upstream package

In this case why don't we update the nixpkgs version instead?

Reply:
crates.io works very well so I don't think a registry needs to be a disaster. In any case,you don't have to use the registry. Of course then you have to type things like `nix build github:NixOS/nixpkgs:...` instead of `nixpkgs:...` all the time.

@timokau Patchelf was just an example, but in general you can't expect Nixpkgs to contain the whole world.

Reply:
Name squatting is also an issue on crates.io, unfortunately. Now, it doesn't have to be a disaster, but it certainly needs moderation and, ideally manual, verification before flake submissions. If the official global registry is small, and users can just add third-party registries to their systems, it shouldn't be bad. Namespacing would also help, but I feel like the official global registry should be limited to official NixOS stuff anyways.

Reply:
> crates.io works very well so I don't think a registry needs to be a disaster. In any case,you don't have to use the registry. Of course then you have to type things like `nix build github:NixOS/nixpkgs:...` instead of `nixpkgs:...` all the time.

You could have the same ergonomics with a local registry only, with the nice commands for adding to it proposed in this rfc, though, with none of the need for us to maintain any sort of global registry.

Reply:
> Ultimately however the monorepo approach fundamentally doesn't scale: it shouldn't be necessary for everybody who just wants to share a NixOS module to have to add it to the nixpkgs repo.

I don’t want to argue this further in this thread, because I think it’s off-topic for the RFC, but I want to make clear that I still don’t agree and any such move would need to be discussed extensively at a later date before it were to happen.

Reply:
> crates.io works very well so I don't think a registry needs to be a disaster. In any case,you don't have to use the registry. Of course then you have to type things like `nix build github:NixOS/nixpkgs:...` instead of `nixpkgs:...` all the time.

You mentioned crates.io a lot in your NixCon talk and it's obvious that this RFC took a lot of inspiration of what Rust is doing. I'd like to chime in here that the public often doesn't see a lot of the effort and sweat that it takes on parts of *people* to run crates.io, and the architectural problems that it comes with. This doesn't just include the actual infrastructure and the team of people it takes to run it, but also a lot of policy decisions on things like DMCA requests, name squatting, etc that introduce a lot of work that general users of the service aren't aware of either. It's easy to say "look, it runs really well" and not see the weekly meetings on policy updates. I got the feeling that everybody involved in developing NixOS was already pretty booked out. I feel that adding yet another thing to maintain to the list of NixOS infrastructure things would be an extremely bad idea.

Reply:
The `crates.io` ecosystem is **fundamentally** a different maintenance approach to NixOS. There are no "Rust maintainers". Sure, there's people who work on the compiler or who are active in various working groups and teams (:wave:) but there's no centralised development in the way that NixOS requires it. Building a bunch of libraries is different from building an operating system distribution and while a global registry of libraries works well for one, it wouldn't scale at all for the other because it's the same people involved in most of the pieces. It just adds overhead.

Reply:
**Re: the mono repo**

As others have said previously, the fact that all the moving parts in NixOS are in a single place, can easily be bisected and tested against one another is a **huge** selling point for collaboration and also lowers the barrier to entry. NixOS is the first Linux distribution I've actually started working on, despite wanting and very much trying to get involved in both Debian and Fedora. I think it would be a shame if this were to change.

There's a relevant discussion/story from the Rust/ `rustc` repo about this. The repo itself is huge, unwieldy, takes ages to download etc. So it was proposed to remove certain components from it that are only tangentially related to the compiler core, such as `rustdoc`. Ultimately, it was decided to keep the repo as it is, because `rustdoc` is a component that frequently breaks with new PRs and so, it's common practise that if a compiler contribution breaks docs, that person then also fixes docs.

I think you could compare the development of `rustc` with the development of `NixOS`/`nixpkgs`. They are both mono repos, both part of an ecosystem with moving pieces around it, but strongly integrated i each other and benefit greatly from having all their parts kept in the same place.